cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,205 @@
+  std::shared_mutex lock_;
+  ServerSessionMiddlewareFactory* factory_;
+  const CallHeaders& headers_;
+  std::shared_ptr<FlightSqlSession> session_;
+  std::string session_id_;
+  const bool existing_session;
+
+ public:
+  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
+                              const CallHeaders& headers)
+      : factory_(factory), headers_(headers), existing_session(false) {}
+
+  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
+                              const CallHeaders& headers,
+                              std::shared_ptr<FlightSqlSession> session,
+                              std::string session_id)
+      : factory_(factory),
+        headers_(headers),
+        session_(std::move(session)),
+        session_id_(std::move(session_id)),
+        existing_session(true) {}
+
+  void SendingHeaders(AddCallHeaders* add_call_headers) override {
+    if (!existing_session && session_) {
+      add_call_headers->AddHeader(
+          "set-cookie", static_cast<std::string>(kSessionCookieName) + "=" + session_id_);
+    }
+  }
+
+  void CallCompleted(const Status&) override {}
+
+  bool HasSession() const override { return static_cast<bool>(session_); }
+
+  std::shared_ptr<FlightSqlSession> GetSession() override {
+    const std::shared_lock<std::shared_mutex> l(lock_);
+    if (!session_) {
+      auto [id, s] = factory_->GetNewSession();
+      session_ = std::move(s);
+      session_id_ = std::move(id);
+    }
+    return session_;
+  }
+
+  const CallHeaders& GetCallHeaders() const override { return headers_; }
+};
+
+std::vector<std::pair<std::string, std::string>>
+ServerSessionMiddlewareFactory::ParseCookieString(const std::string_view& s) {
+  const std::string list_sep = "; ";
+  const std::string pair_sep = "=";
+  const size_t pair_sep_len = pair_sep.length();
+
+  std::vector<std::pair<std::string, std::string>> result;
+
+  size_t cur = 0;
+  while (cur < s.length()) {
+    const size_t end = s.find(list_sep, cur);
+    size_t len;
+    if (end == std::string::npos) {
+      // No (further) list delimiters
+      len = std::string::npos;
+      cur = s.length();
+    } else {
+      len = end - cur;
+      cur = end;
+    }
+    const std::string_view tok = s.substr(cur, len);
+
+    const size_t val_pos = tok.find(pair_sep);
+    if (val_pos == std::string::npos) {
+      // The cookie header is somewhat malformed; ignore the key and continue parsing
+      continue;
+    }
+    result.emplace_back(tok.substr(0, val_pos),
+                        tok.substr(val_pos + pair_sep_len, std::string::npos));
+  }
+
+  return result;
+}
+
+Status ServerSessionMiddlewareFactory::StartCall(
+    const CallInfo&, const CallHeaders& incoming_headers,
+    std::shared_ptr<ServerMiddleware>* middleware) {
+  std::string session_id;
+
+  const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&
+      headers_it_pr = incoming_headers.equal_range("cookie");
```

Comment:
Is this case-sensitive?

Reply:
gRPC only ever uses lowercase headers (the c++ impl will actually crash if you don't)

Reply:
Oh, do you mean gRPC will auto-lower-case incoming headers?

Reply:
It should, and apparently, this extends down to HTTP/2 as well